integration-tests/plugins/__tests__/workflows/product/create-product.ts

```diff
@@ -0,0 +1,136 @@
+import path from "path"
+import { initDb, useDb } from "../../../../environment-helpers/use-db"
+import { bootstrapApp } from "../../../../environment-helpers/bootstrap-app"
+import { Actions, createProducts, InputAlias, pipe } from "@medusajs/workflows"
+import { WorkflowTypes } from "@medusajs/types"
+import {
+  defaultAdminProductFields,
+  defaultAdminProductRelations,
+} from "@medusajs/medusa/dist"
+import { listProducts } from "@medusajs/workflows/dist/handlers/product"
+
+describe("CreateProduct workflow", function () {
+  let medusaProcess
+  let medusaContainer
+
+  beforeAll(async () => {
+    const cwd = path.resolve(path.join(__dirname, "..", "..", ".."))
+    await initDb({ cwd } as any)
+    const { container } = await bootstrapApp({ cwd })
+    medusaContainer = container
+  })
+
+  afterAll(async () => {
+    const db = useDb()
+    await db.shutdown()
+
+    medusaProcess.kill()
+  })
+
+  it("should compensate all the invoke if something fails", async () => {
+    const workflow = createProducts(medusaContainer)
+
+    workflow.appendAction(
+      "fail_step",
+      Actions.result,
+      {
+        invoke: pipe(
+          {
+            inputAlias: InputAlias.ProductsInputData,
+            invoke: {
+              from: InputAlias.ProductsInputData,
+            },
+          },
+          async function failStep() {
+            throw new Error(`Failed to create products`)
+          }
+        ),
+      },
+      {
+        noCompensation: true,
+      }
+    )
+
+    const input: WorkflowTypes.ProductWorkflow.CreateProductsWorkflowInputDTO =
+      {
+        products: [
+          {
+            title: "Test product",
+            type: { value: "physical" },
+            tags: [{ value: "test" }],
+            subtitle: "Test subtitle",
+            variants: [
+              {
+                title: "Test variant",
+                prices: [
+                  {
+                    amount: 100,
+                    currency_code: "usd",
+                  },
+                ],
+              },
+            ],
+            options: [
+              {
+                title: "Test option",
+              },
+            ],
+          },
+        ],
+        config: {
+          listConfig: {
+            select: defaultAdminProductFields,
+            relations: defaultAdminProductRelations,
+          },
+        },
+      }
+
+    const manager = medusaContainer.resolve("manager")
+    const context = {
+      manager,
+    }
+    const { result, errors, transaction } = await workflow.run({
+      input,
+      context,
+      throwOnError: false,
+    })
+
+    expect(errors).toEqual([
+      {
+        action: "fail_step",
+        handlerType: "invoke",
+        error: new Error(`Failed to create products`),
+      },
+    ])
```

Comment:
since data in the compensation steps are really important to inspect, should we test the data saved in each compensation steps?

Reply:
what do you mean? most of the method return void, here when I delete the product from the product module, I trust the product module (which is tested) to also cascade the other entities for example

Reply:
so I don't have data saved in the compensation steps

Reply:
Maybe I ve misunderstood what you mean, let me know, but here the preliminary objective of that test is to ensure that at the end of the compensation we are in the right state of the data. A bit like when you test an endpoint, you check what data you got at the end of the entire floe

Reply:
I don't think the integration tests is the place for that, shouldn't it be part of the unit tests where we will check the haveBeenCalledWith etc?

Reply:
looks like jest on my side fail to run the following for some reason, it just run indefinitely, @carlos-r-l-rodrigues @riqwan any idea?
```ts
expect(transactionContext).toEqual(
      expect.objectContaining({
        invoke: expect.objectContaining({
          [CreateProductsActions.prepare]: {
            products: [
              expect.objectContaining({
                title: input.products[0].title,
                type: input.products[0].type,
                tags: input.products[0].tags,
                subtitle: input.products[0].subtitle,
                variants: input.products[0].variants,
                options: input.products[0].options,
                handle: kebabCase(input.products[0].title),
              }),
            ],
            productsHandleSalesChannelsMap: expect.any(Object),
            productsHandleVariantsIndexPricesMap: expect.any(Object),
            config: input.config,
          },
        }),
        [CreateProductsActions.createProducts]: expect.objectContaining({
          id: expect.any(String),
          title: input.products[0].title,
          type: {
            id: expect.any(String),
          },
          variants: [
            {
              id: expect.any(String),
            },
          ],
          options: [
            {
              id: expect.any(String),
            },
          ],
          tags: [
            {
              id: expect.any(String),
            },
          ],
        }),
        [CreateProductsActions.createPrices]: undefined,
        [CreateProductsActions.createInventoryItems]: [
          {
            variant: expect.objectContaining({
              id: expect.any(String),
            }),
            inventoryItem: expect.objectContaining({
              id: expect.any(String),
            }),
          },
        ],
        [CreateProductsActions.attachInventoryItems]: [
          expect.objectContaining({
            inventory_item_id: expect.any(String),
            variant_id: expect.any(String),
            required_quantity: 1,
            id: expect.any(String),
          }),
        ],
      })
    )
```

Reply:
Ah crap sorry just saw this. I do believe we have a misunderstanding. Thanks for trying,  we can pick this up later. It was just a thought for now.